### ISSUE TITLE
Remove need for a miniconda installation ?

### ISSUE THREAD

**Issue by Luthaf** [Luthaf/Conda.jl#1]
It could be feasible to parse directly the metadata in the `info` folder to resolve the dependencies directly in Julia. 

Pro:
 - Remove the need for a full miniconda installation, and minimize internet usage;
 - (Maybe) faster operations, as their is no need to shell out commands, and only the needed code will be used;

Con:
 - Why rewrite the wheel?
 - Writing a correct package manager is hard.

**Comment by dhoegh** [comment]
I use your package because it install the miniconda and therefore I do not think you should remove it. If you do could you make a new package? My purpose is to make my Julia work environment easy to install even when I depend on both matplotlib and IPython notebooks. Your package fulfill this purpose perfectly as my installation script is:
```
Pkg.add("Conda")
using Conda
# Installing the notebook interface
Conda.add("ipython-notebook")
# Installing the plotting library
Conda.add("matplotlib")
#################### restart terminal ####################
# installs notebook
Pkg.add("IJulia")
# Plotting library
Pkg.add("PyPlot")
```
This is a lot easier to tell people to do than, asking them to install the Miniconda manually and using cmd to manage packages and the space required is smaller than using the whole Anaconda distribution.

**Comment by Luthaf** [comment]
Removing miniconda will not prevent you from doing this. The same packages would be present, and if a Python distribution is needed by a package, it would be installed. 

```julia
Conda.add("matplotlib")
```

Would still download and install Python, Numpy, matplotlib and any dependency.

The idea here is to remove this python distribution when it is not used (i.e. when managing binary dependencies for BinDeps). But as this issue would require a pretty big amount of work (writing a correct package manager is hard), and my time is limited, Miniconda is not going away anytime soon.

**Comment by dhoegh** [comment]
I see, good to know. I am promoting your package as the easiest way of getting IJulia installed.

**Comment by dpsanders** [comment]
@dhoegh Maybe you know that now all you need to do is
```julia
Pkg.add("IJulia")
```
and that will automatically install `Conda` etc.

**Comment by dhoegh** [comment]
Yes, I pushed the use of Conda in IJulia, PyCall and PyPlot forward.

**Comment by dpsanders** [comment]
OK, thanks very much for doing that!

**Comment by jakirkham** [comment]
I would like to add one more con here. Namely, you are now maintaining your own version of `conda`, which has the potential to diverge with time. This becomes more important when you want to deal with packages that are `conda` maintained.

**Comment by cortner** [comment]
If I already have the full Anaconda distribution, then `Conda.jl` still tries to install miniconda:
```
julia> Conda.add("pyamg")
INFO: Downloading miniconda installer ...
```
Is this intended behaviour - why does it not simply use the existing installation?  If I let this go ahead, will it mess up my existing anaconda installation?

Thanks.

**Comment by Luthaf** [comment]
Yes, this is the intended behaviour: Conda.jl is using it's own version of miniconda exactly to prevent it from messing up any existing anaconda installation (and to prevent the user from messing up the Conda.jl installation ^^).

I really should a

**Comment by cortner** [comment]
It would also be really useful if there could be some examples how to setup `build.jl` to choose install a Python dependency correctly, i.e., using the existing installation if it can be found and `Conda.jl` otherwise?

**Comment by hayd** [comment]
@Luthaf  IJulia.jl seems to have a slightly different policy (it uses the global, potentially incomplete, ipython i.e. even if it's without jupyter) https://github.com/JuliaLang/IJulia.jl/issues/363#issuecomment-220792927

**Comment by Luthaf** [comment]
> It would also be really useful if there could be some examples how to setup build.jl to choose install a Python dependency correctly, i.e., using the existing installation if it can be found and Conda.jl otherwise?

@stevengj just added functionality for that in PyCall. See https://github.com/JuliaLang/METADATA.jl/pull/5201 and links for the API.

> @Luthaf IJulia.jl seems to have a slightly different policy (it uses the global, potentially incomplete, ipython i.e. even if it's without jupyter) JuliaLang/IJulia.jl#363 (comment)

As I wrote it, `Conda.jl` provide access to `conda`, not `python`. The fact that `conda` is written in Python is purely an implementation detail. 

The main goal of Conda.jl is to provide a cross-platform access to native (C, C++ and Fortran) dependencies, not Python dependencies =). So it is completely up to users of Conda.jl to decide what to do with it.

The IJulia.jl installation error should have it's own issue. I think this is some kind of bad interaction between packages.

**Comment by cortner** [comment]
Yes - I looked at PyCall. the file is about 300 lines, and given enough time I could probably dissect it. I am more looking for a 
```
if has_anaconda_installation()
   run(`conda install somepackage`)
else 
  Conda.add("somepackage")
end
```

**Comment by Luthaf** [comment]
I was more thinking about the `pyimport_conda("<pip name>", "<conda name>")` function from PyCall. It tries to import the python package from the first argument, and if that fails and PyCall is using Conda, then it installs the package. Else it displays a message to the user giving instructions about how to use PyCall with Conda.jl.

But you could try to implement `has_anaconda_installation()` like this:
```julia
try
    run(`conda install somepackage`)
catch
    Conda.add("somepackage")
end
```

Except it might not work all the time. If you are trying to get a Python dependency, I strongly recommend you use the `PyCall.pyimport_conda` function.

**Comment by cortner** [comment]
thanks. so I could go the other way round, use 
```
try 
     pyimport_conda("<pip name>", "<conda name>")
catch
    run(`conda install some package`)
end
```

**Comment by stevengj** [comment]
@cortner, the `pyimport_conda` function does the `conda install` for you.  That's the whole point.

Note that this is documented in the PyCall README ... no need to read the PyCall source code.

**Comment by stevengj** [comment]
(Note that the first argument of `pyimport_conda` is the name of the module you want to import; it has nothing to do with pip.)

**Comment by cortner** [comment]
@stevengj Thank you and sorry for missing that.

**Comment by cortner** [comment]
@stevengj I tried to use`pyimport_conda`, but without success:

```
julia> Pkg.build("PyCall")
INFO: Building PyCall
INFO: PyCall is using /Users/ortner/anaconda/bin/python (Python 2.7.11) at /Users/ortner/anaconda/bin/python, libpython = /Users/ortner/anaconda/lib/libpython2.7.dylib

julia> PyCall.pyimport_conda("pyamg", "pyamg")
ERROR: Failed to import required Python module pyamg.

For automated pyamg installation, try configuring PyCall to use the Conda Python distribution within Julia.  Relaunch Julia and run:
    ENV["PYTHON"]=""
    Pkg.build("PyCall")
before trying again.
#  and some more
```

But I specifically don't want to use the miniconda distribution. I also tried it with first setting
```
julia> ENV["PYTHON"] = "/Users/ortner/anaconda/bin/python"
"/Users/ortner/anaconda/bin/python"
```

Where is my mistake? Thank you.

**Comment by Luthaf** [comment]
@cortner Could you please open a separated issue for your problem? This one is for discussing the implementation of the `conda` package manager to use. Thanks!

**Comment by manslogic** [comment]
Hi,I am new to Julia,
Please help me to get around this problem, i am trying to install **jupyter notebook**,
i am facing below problem.

![image](https://user-images.githubusercontent.com/1926423/54581537-28cfb800-4a33-11e9-8625-782274ce210a.png)


log:
````
		julia> Conda.add("ipython-notebook")
		[ Info: Downloading miniconda installer ...
		ERROR: failed process: Process(`'C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe' -Version 3 -NoProfile -Command "[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12; (New-Object System.Net.Webclient).DownloadFile('https://repo.continuum.io/miniconda/Miniconda3-4.5.4-Windows-x86_64.exe', 'C:\Users\trinity\.julia\conda\3\installer.exe')"`, ProcessExited(3221225477)) [3221225477]
		Stacktrace:
		[1] error(::String, ::Base.Process, ::String, ::Int64, ::String) at .\error.jl:42
		[2] pipeline_error at .\process.jl:785 [inlined]
		[3] download(::String, ::String) at .\download.jl:20
		[4] _install_conda(::String, ::Bool) at C:\Users\trinity\.julia\packages\Conda\CpuvI\src\Conda.jl:160
		[5] _install_conda(::String) at C:\Users\trinity\.julia\packages\Conda\CpuvI\src\Conda.jl:152
		[6] runconda(::Cmd, ::String) at C:\Users\trinity\.julia\packages\Conda\CpuvI\src\Conda.jl:111
		[7] add at C:\Users\trinity\.julia\packages\Conda\CpuvI\src\Conda.jl:183 [inlined] (repeats 2 times)
		[8] top-level scope at none:0

````

**Comment by stevengj** [comment]
@manslogic, please don't post help requests to unrelated github issues.  You are having a download problem; that might indicate a firewall or something on your machine.  You could try asking for help on https://discourse.julialang.org/
